sdk/core/core-https/src/nodeHttpsClient.ts

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as https from "https";
+import * as zlib from "zlib";
+import { Transform } from "stream";
+import FormData from "form-data";
+import httpsProxyAgent from "https-proxy-agent";
+import { AbortController, AbortError } from "@azure/abort-controller";
+import {
+  HttpsClient,
+  PipelineRequest,
+  PipelineResponse,
+  TransferProgressEvent,
+  HttpHeaders,
+  FormDataMap
+} from "./interfaces";
+import { createHttpHeaders } from "./httpHeaders";
+import { RestError } from "./restError";
+import { URL } from "./util/url";
+import { IncomingMessage } from "http";
+
+let keepAliveAgent: https.Agent;
+let proxyAgent: https.Agent;
+
+function getOrCreateAgent(request: PipelineRequest): https.Agent {
+  // At the moment, proxy settings and keepAlive are mutually
+  // exclusive because the proxy library currently lacks the
+  // ability to create a proxy with keepAlive turned on.
+  const proxySettings = request.proxySettings;
+  if (proxySettings) {
+    if (!proxyAgent) {
+      const proxyAgentOptions: httpsProxyAgent.HttpsProxyAgentOptions = {
+        host: proxySettings.host,
+        port: proxySettings.port,
+        headers: request.headers.toJSON()
+      };
+      if (proxySettings.username && proxySettings.password) {
+        proxyAgentOptions.auth = `${proxySettings.username}:${proxySettings.password}`;
+      }
+      proxyAgent = new httpsProxyAgent(proxyAgentOptions);
+    }
+    return proxyAgent;
+  } else if (request.keepAlive) {
+    if (!keepAliveAgent) {
+      keepAliveAgent = new https.Agent({
+        keepAlive: true
+      });
+    }
+
+    return keepAliveAgent;
+  } else {
+    return https.globalAgent;
+  }
+}
+
+function isReadableStream(body: any): body is NodeJS.ReadableStream {
+  return body && typeof body.pipe === "function";
+}
+
+class ReportTransform extends Transform {
+  private loadedBytes = 0;
+  private progressCallback: (progress: TransferProgressEvent) => void;
+  _transform(chunk: string | Buffer, _encoding: string, callback: Function): void {
+    this.push(chunk);
+    this.loadedBytes += chunk.length;
+    this.progressCallback({ loadedBytes: this.loadedBytes });
```

Comment:
put a finally here to make sure we continue the transform, but I'm not going to swallow the exception unless you think that's a good idea

Reply:
I'm not sure I get what you mean with the finally, but does it make sense then to call `callback()` with the error thrown from `progressCallback`? (Then we'd also need to propagate the `error` emitted from `ReportTransform` streams such that they can catch it from the operation they called)

Reply:
oh I see, you want to pipe the error into the callback so the stream will raise an error? That works, since we listen to error when we convert it to text and they should listen if they use the stream directly

Reply:
Fixed this and added a test